src/services/connection.service.ts

```diff
@@ -0,0 +1,221 @@
+                }
+            }
+
+            const implementation = this.receiveHandlers.get(message.type);
+            if (implementation && typeof implementation === "function") {
+                await implementation(message, message.originalSenderGuid, (responseMessage: Message) => {
+                    responseMessage.correlationId = message.guid;
+                    responseMessage.originalSenderGuid = message.originalSenderGuid;
+                    this.sendMessage(responseMessage, responseMessage.originalSenderGuid);
+                });
+                // Acknowledge this message
+                if (message.type !== MessageType.ACKNOWLEDGE) {
+                    const destination = this.getIpFromRoutingTable(message.originalSenderGuid);
+                    this.sender.sendAcknowledgeMessage(message, destination);
+                }
+            }
+            resolve();
+        });
+    }
+
+    /**
+     * Leave the network.
+     */
+    public leave(guid: string): void {
+        const message = new Message(MessageType.LEAVE, guid);
+        this.sendBroadcast(message);
+    }
+
+
+    public sendMessageByIp(message: Message, destinationIp: string, responseImplementation?: RESPONSE_TYPE): Promise<void> {
+        return new Promise(async (resolve, reject) => {
+            if (!this.sender) {
+                reject();
+                return;
+            }
+            
+            if (responseImplementation) {
+                this.requestsMap.set(message.guid, responseImplementation);
+                this.responseDefferedsMap.set(message.guid, new Deferred());
+            }
+            await this.sender.sendMessage(message, destinationIp);
+            resolve();
+        });
+    }
+
+    public getPromiseForResponse(message: Message): Promise<void> {
+        return new Promise(async (resolve, reject) => {
+            const responseDeffered = this.responseDefferedsMap.get(message.correlationId);
+            if (responseDeffered) {
+                await responseDeffered.promise;
+                resolve();
+            }
+            reject();
+        });
+    }
+
+    /**
+     * Get guids from the senders of the sent messages after the given date.
+     * Deletes messages that are sent before the given date from the history.
+     *
+     * @param date - The date
+     *
+     * @return An array of the GUIDs of the senders
+     */
+    private getSentMessagesSendersSince(date: Date): string[] {
+        const guids: string[] = [];
+
+        for (const guid of this.sentMessages.keys()) {
+            const message = this.sentMessages.get(guid);
+
+            if (message && message.isOlderThan(date) && guid) {
+                guids.push(guid);
+            }
+        }
+        return guids;
+    }
+
+
+    /**
+     * Create timer that removes peers that did not reply to a sent message.
+     *
+     * Used to remove offline peers
+     */
+    private createRoutingTableCleanupTimer() {
```

Comment:
Please consider using the `SchedulableJob` here that we're using in the `Blockr validator`. 
Please note that if you'd implement the `SchedulableJob`, we do not want duplicate code so you'll have to figure out how to go around that.